src/puppetlabs/puppetdb/anonymizer.clj

```diff
@@ -304,109 +298,89 @@
-   context
-   config]
+        (update "message" anonymize-leaf :message newcontext config)
+        (update "property" anonymize-leaf :parameter-name newcontext config)
+        (update "new_value" anonymize-leaf :parameter-value (assoc newcontext :parameter-value new_value) config)
+        (update "old_value" anonymize-leaf :parameter-value (assoc newcontext :parameter-value old_value) config))))
+
+(defn anonymize-events [events context config]
+  (map #(anonymize-event % context config) events))
+
+(defn anonymize-report-resource
+  [resource context config]
+  (let [newcontext {"node" (get context "node")
+                    "title" (get resource "resource_title")
+                    "type" (get resource "resource_type")
+                    "file" (get resource "file")
+                    "line" (get resource "line")}]
+    (-> resource
+        (update "resource_title" anonymize-leaf :title newcontext config)
+        (update "resource_type" anonymize-leaf :type newcontext config)
+        (utils/update-when ["file"] anonymize-leaf :file newcontext config)
+        (utils/update-when ["line"] anonymize-leaf :line newcontext config)
+        (utils/update-when ["containment_path"] anonymize-containment-path newcontext config)
+        (update "events" anonymize-events newcontext config))))
+
+(pls/defn-validated anonymize-report-resources :- [resource-schema-str]
+  [resources :- [resource-schema-str]
+   context config]
+  (map #(anonymize-report-resource % context config) resources))
+
+(defn anonymize-metric [metric context config]
   (if (= "time" (get metric "category"))
-    (update-in metric ["name"] #(anonymize-lowercase-type % context config))
+    (update metric "name" #(anonymize-lowercase-type % context config))
     metric))
 
 (pls/defn-validated anonymize-metrics :- [metric-schema-str]
-  [metrics :- (s/maybe [metric-schema-str])
+  [metrics :- [metric-schema-str]
    context
    config]
-  (when metrics
-    (map #(anonymize-metric % context config) metrics)))
+  (map #(anonymize-metric % context config) metrics))
 
-(pls/defn-validated anonymize-log :- log-schema-str
-  [log :- log-schema-str
-   context
-   config]
+(defn anonymize-log [log context config]
   (-> log
-      (update-in ["message"] anonymize-leaf :log-message context config)
-      (update-in ["source"] anonymize-log-source context config)
-      (update-in ["tags"] anonymize-tags context config)
-      (update-in ["file"] anonymize-leaf :file context config)
-      (update-in ["line"] anonymize-leaf :line context config)))
+      (update "message" anonymize-leaf :log-message context config)
+      (update "source" anonymize-log-source context config)
+      (update "tags" anonymize-tags context config)
+      (update "file" anonymize-leaf :file context config)
+      (update "line" anonymize-leaf :line context config)))
 
-(pls/defn-validated anonymize-logs :- (s/maybe [log-schema-str])
+(pls/defn-validated anonymize-logs :- [log-schema-str]
```

Comment:
You know, I didn't actually believe you about schema letting `nil` through here. But I checked, and that's what it does. I think it's to do with the strangeness of the way the clojure sequence functions handle `nil` in general. I suppose it's common to have odd behavior around function zeros, but it's still unsettling. 

```clojure
(count nil) ; 0
(count []) ; 0
(nth nil 0) ; nil
(nth [] 0) ; => exception
(vec nil) ; []
```